relays/ethereum/src/message_race_delivery.rs

```diff
@@ -135,34 +136,60 @@ where
 	}
 }
 
-/// Message delivery strategy.
-struct MessageDeliveryStrategy<P: MessageLane> {
+/// Messages delivery strategy.
+type MessageDeliveryStrategy<P> = DeliveryStrategy<
+	<P as MessageLane>::SourceHeaderNumber,
+	<P as MessageLane>::SourceHeaderHash,
+	<P as MessageLane>::TargetHeaderNumber,
+	<P as MessageLane>::TargetHeaderHash,
+	<P as MessageLane>::MessageNonce,
+	<P as MessageLane>::MessagesProof,
+>;
+
+/// Nonces delivery strategy.
+#[derive(Debug)]
+pub struct DeliveryStrategy<SourceHeaderNumber, SourceHeaderHash, TargetHeaderNumber, TargetHeaderHash, Nonce, Proof> {
```

Comment:
Why do we split the generic parameters instead of simply using `<P: MessageLane>`? The extra bounds could be added via `where` clause, no?

Reply:
Because when `DeliveryStrategy` is used in message-delivery-race, then the `race.source = lane.source && race.target = lane.target`. When used in delivery-confirmations-race, then the `race.source = lane.target && race.target = lane.source`.

Reply:
Okay, fair! We could implement `struct FlipSourceTarget<P: MessageLane>; impl MessageLane for FlipSourceTarget<P> {}` though, couldn't we and then just wrap it?

Reply:
This would work, but `MessageLane` is not only about source/target chain headers hashes/numbers. So it doesn't seems right.